service-mirror.go

```diff
@@ -6,19 +6,330 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"reflect"
 	"strings"
+	"time"
 
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
 	apiError "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/cache"
+	client "k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/util/homedir"
 )
 
+// Make sure all global services lies in only one namespace.
+const GLOBAL_SVC_NAMESPACE = "default"
```

Comment:
Hmmm this is good, but we should add a flag for it maybe. The services should use the same namespace as the controller. When we run locally, we'd expect the namespace to be `""`. When we deploy, we'd expect it to be the same namespace as the controller. Here's an example:

```
# Run locally, don't pass in a namespace. Default namespace in k8s is also "".
go run ./main.go 
```

```
# A hypothetical deployment template spec:
- name: rushi-controller:v0.0.1
  args:
  - --controller-namespace
  - "global-prototype" # in helm we can template this {{ .Values.namespace }}
 ```

Reply:
Yess I agree with Flag but i have doubt regarding running controller in same namespace as spawning global-svc. I am smh kind biased towards putting it in default as other services will be mirrored across different namespace. As global service is aggregator will put it in default but i agree that this is should be configurable.

Regarding `""` in using namespace, smh client-go was giving me issue when I was creating service. Error was kind of : if you creating or giving resource name, namespace couldn't be empty

Reply:
Hmm, you're right that the controller should put it in its own namespace. Perhaps we don't need a flag for this at all. What if instead we just adapt to the services we are mirroring?

Let's take two examples:

```
msql service mirrors from target1 and target2

. mysql namespace
|___ mysql (service created by source)
|___ mysql-target1
|___mysql-target2

redis service mirror from target1

.redis namespace
|___ redis
|___redis-target1
```

Let's ignore for a while that someone would use both 😅 . We can see that people would expect the mirror to be in the _actual_ namespace as the service they have. Typically, applications will live in the same namespace too. We should do the same and add the global mirror in whatever namespace the rest of the services are deployed in.

Namespaces typically mean ownership in k8s, and they have a few other implications:
1. calling services is easier (you don't need the whole fqdn, just the service name and port)
2. routing and policy are namespace dependent; dynamic routing and policy for example takes its scope depending on which namespace the routes are declared in.

Typically a "service owner" (for a database it would be the platform team) manage their service and give some routing defaults. Other clients are free to consume the service and provide their own policy on top of it. 

Let's merge this as is now (default namespace) and create an issue for it.

Reply:
Thanks for the explanation matei. Well to add to your last point so i have already added flag if its not provided its default to default 😅 . Else it uses the namespace provided. But with your first point that make sense. Probably later we can make both cases are handled.